csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs

```diff
@@ -78,23 +79,24 @@ protected async ValueTask<RecordBatch> ReadRecordBatchAsync(CancellationToken ca
             }
 
             RecordBatch result = null;
-            await Buffers.RentReturnAsync(messageLength, async (messageBuff) =>
+            await ArrayPool<byte>.Shared.RentReturnAsync(messageLength, async (messageBuff) =>
             {
                 int bytesRead = await BaseStream.ReadFullBufferAsync(messageBuff, cancellationToken)
                     .ConfigureAwait(false);
                 EnsureFullRead(messageBuff, bytesRead);
 
                 Flatbuf.Message message = Flatbuf.Message.GetRootAsMessage(CreateByteBuffer(messageBuff));
 
-                await Buffers.RentReturnAsync((int)message.BodyLength, async (bodyBuff) =>
-                {
-                    int bodyBytesRead = await BaseStream.ReadFullBufferAsync(bodyBuff, cancellationToken)
-                        .ConfigureAwait(false);
-                    EnsureFullRead(bodyBuff, bodyBytesRead);
+                int bodyLength = checked((int)message.BodyLength);
+
+                IMemoryOwner<byte> bodyBuffOwner = _allocator.Allocate(bodyLength);
+                Memory<byte> bodyBuff = bodyBuffOwner?.Memory.Slice(0, bodyLength) ?? Memory<byte>.Empty;
+                bytesRead = await BaseStream.ReadFullBufferAsync(bodyBuff, cancellationToken)
+                    .ConfigureAwait(false);
+                EnsureFullRead(bodyBuff, bytesRead);
 
-                    FlatBuffers.ByteBuffer bodybb = CreateByteBuffer(bodyBuff);
-                    result = CreateArrowObjectFromMessage(message, bodybb);
-                }).ConfigureAwait(false);
+                FlatBuffers.ByteBuffer bodybb = CreateByteBuffer(bodyBuff);
+                result = CreateArrowObjectFromMessage(message, bodybb, bodyBuffOwner);
```

Comment:
should the owner be disposed here?

edit: I can see that `RecordBatch` is disposable now. Is the Arrow deserialization lazy and we need to keep the memory alive until somebody reads from it?

Reply:
The deserialization happens during `CreateArrowObjectFromMessage`. That method creates a `RecordBatch` and associated `IArrowArray` objects pointing into the memory here. The memory needs to be kept alive because how else would we remember all the values that we just read from the Stream?

Reply:
It makes sense to me now #solved